docs/en_US/NAS/Serialization.rst

```diff
@@ -0,0 +1,52 @@
+Serialization
+=============
+
+In multi-trial NAS, a sampled model should be able to be executed on a remote machine or a training platform (e.g., AzureML, OpenPAI). "Serialization" enables re-instantiation of model evaluator in another process or machine, such that, both the model and its model evaluator should be correctly serialized. To make NNI correctly serialize model evaluator, users should apply ``nni.trace`` on some of their functions and objects. API references can be found in :func:`nni.trace`.
+
+Serialization is implemented as a combination of `json-tricks <https://json-tricks.readthedocs.io/en/latest/>`_ and `cloudpickle <https://github.com/cloudpipe/cloudpickle>`_. Essentially, it is json-tricks, that is a enhanced version of Python JSON, enabling handling of serialization of numpy arrays, date/times, decimal, fraction and etc. The difference lies in the handling of class instances. Json-tricks deals with class instances with ``__dict__`` and ``__class__``, which in most of our cases are not reliable (e.g., datasets, dataloaders). Rather, our serialization deals with class instances with two methods:
+
+1. If the class / factory that creates the object is decorated with ``nni.trace``, we can serialize the type / factory function, along with the parameters, such that the instance can be re-instantiated.
+2. Otherwise, cloudpickle is used to serialize the object into a binary.
+
+The recommendation is, unless you are absolutely certain that there is no problem and extra burden to serialize the object into binary, always add ``nni.trace``. In most cases, it will be more clean and neat, and enables possibilities such as mutation of parameters (will be supported in future).
+
+.. warning::
+
+    **What will happen if I forget to "trace" my objects?**
+
+    It is likely that the program can still run. NNI will try to serialize the untraced object into a binary. If might fail in complicated cases (e.g., circular dependency). Even if it succeeds, the result might be a substantially large object. For example, if you forgot to add ``nni.trace`` on ``MNIST``, the MNIST dataset object wil be serialized into binary, which will be dozens of megabytes because the object has the whole 60k images stored inside. You might see warnings and even errors when running experiments. To avoid such issues, the easiest way is to always remember to add ``nni.trace`` to non-primitive objects.
+
+To trace a function or class, users can use decorator like,
+
+.. code-block:: python
+
+    @nni.trace
+    class MyClass:
+        ...
+
+Inline trace that traces instantly before being used is also acceptable: ``nni.trace(MyClass)(parameters)``.
```

Comment:
what will happen if I both decorate class with `@nni.trace` and use `nni.trace()` on instantiation?

Reply:
The second time `nni.trace` is called, it will detect that it is already `traceable`, so it does nothing.